src/ray/raylet/node_manager.cc

```diff
@@ -2766,6 +2769,91 @@ void NodeManager::SendSpilledObjectRestorationRequestToRemoteNode(
+    }
+
+    if (!resources_changes->empty()) {
+      for (const auto &iter : *resources_changes) {
+        (*resources_data->mutable_normal_task_resources_changes())[iter.first] =
+            iter.second;
+      }
+      last_report_normal_task_resources_ = new_resources;
+    }
+  }
+}
+
+std::shared_ptr<std::unordered_map<std::string, double>>
+NodeManager::GetResourcesUsedByNormalTask() const {
+  auto resources = std::make_shared<std::unordered_map<std::string, double>>();
+  if (!RayConfig::instance().new_scheduler_enabled()) {
+    return resources;
+  }
+
+  auto cluster_resource_scheduler =
+      std::dynamic_pointer_cast<ClusterResourceScheduler>(cluster_resource_scheduler_);
+  for (const auto &iter : leased_workers_) {
+    const auto &leased_worker = iter.second;
+    const auto &allocated_instances = leased_worker->GetAllocatedInstances();
+    if (leased_worker->GetActorId().IsNil() && allocated_instances != nullptr) {
+      const auto &predefined_resources = allocated_instances->predefined_resources;
+      for (size_t res_idx = 0; res_idx < predefined_resources.size(); res_idx++) {
+        for (size_t inst_idx = 0; inst_idx < predefined_resources[res_idx].size();
+             inst_idx++) {
+          if (predefined_resources[res_idx][inst_idx] > 0.) {
+            const auto &resource_name =
+                cluster_resource_scheduler->GetResourceNameFromIndex(res_idx);
+            (*resources)[resource_name] +=
+                predefined_resources[res_idx][inst_idx].Double();
+          }
+        }
+      }
+      auto custom_resources = allocated_instances->custom_resources;
+      for (auto it = custom_resources.begin(); it != custom_resources.end(); ++it) {
+        for (size_t inst_idx = 0; inst_idx < it->second.size(); inst_idx++) {
+          if (it->second[inst_idx] > 0.) {
+            const auto &resource_name =
+                cluster_resource_scheduler->GetResourceNameFromIndex(it->first);
+            (*resources)[resource_name] += it->second[inst_idx].Double();
+          }
+        }
+      }
+    }
+  }
+  return resources;
+}
+
+std::shared_ptr<std::unordered_map<std::string, double>>
+NodeManager::GetNormalTaskResourcesChanges(
+    std::unordered_map<std::string, double> old_resources,
+    const std::shared_ptr<std::unordered_map<std::string, double>> &new_resources) const {
+  auto resources_changes = std::make_shared<std::unordered_map<std::string, double>>();
+  for (const auto &resource : *new_resources) {
+    const auto &iter = old_resources.find(resource.first);
+    if (iter == old_resources.end()) {
+      (*resources_changes)[resource.first] = resource.second;
+    } else {
+      old_resources.erase(iter);
+      if (fabs(resource.second - iter->second) > 1e-6) {
+        (*resources_changes)[resource.first] = resource.second - iter->second;
+      }
+    }
+  }
+
+  for (const auto &resource : old_resources) {
+    (*resources_changes)[resource.first] = -resource.second;
```

Comment:
Why do we subtract old resources again?

Reply:
It reports the change of resource used by task. If the task (using resource A) finish, `new_resources` does not contain resource A, so we need to obtain resource A from `old_resources` and report it to GCS, indicating that resource A has been returned.

Reply:
Can you comment this?

Reply:
I will add comment, thx.